File: jscomp/test/gpr_1072.ml
(* external ice_cream: ?flavor:([`vanilla | `chocolate ] [@bs.string]) ->
   num:int -> unit -> _ = "" [@@bs.obj]

   let my_scoop = ice_cream ~flavor:`vanilla ~num:3 () *)
(* external ice_cream_2: flavor:([`vanilla | `chocolate ] [@bs.string]) ->
   num:int -> unit -> _ = "" [@@bs.obj]

   let my_scoop2 = ice_cream_2 ~flavor:`vanilla ~num:3 () *)

type opt_test = < x: int Js.Undefined.t ; y: int Js.Undefined.t > Js.t

external opt_test : ?x:int -> ?y:int -> unit -> _ = "" [@@bs.obj]

let u : opt_test = opt_test ~y:3 ()

external ice_cream3 :
     ?flavor:([`vanilla | `chocolate[@bs.as "x"]][@bs.string])
  -> num:int
  -> unit
  -> _ = ""
  [@@bs.obj]

(* TODO: warn when [_] happens in any place except `bs.obj` *)

type ice_cream3_expect = < flavor: string Js.undefined ; num: int > Js.t

let v_ice_cream3 : ice_cream3_expect list =
  [ ice_cream3 ~flavor:`vanilla ~num:3 ()
  ; ice_cream3 ~flavor:`chocolate ~num:3 ()
  ; ice_cream3 ~flavor:`vanilla ~num:3 () ]

type u

external ice_cream4 :
     ?flavor:([`vanilla | `chocolate[@bs.as "x"]][@bs.string])
  -> num:int
  -> unit
  -> u = ""
  [@@bs.obj]

let v_ice_cream4 : u list =
  [ ice_cream4 ~flavor:`vanilla ~num:3 ()
  ; ice_cream4 ~flavor:`chocolate ~num:3 () ]

external label_test : x__ignore:int -> unit -> _ = "" [@@bs.obj]

type label_expect = < x__ignore: int > Js.t
(** here the type label should be the same, when get the object, it will be
    mangled *)

let vv : label_expect = label_test ~x__ignore:3 ()

external int_test : x__ignore:([`a | `b][@bs.int]) -> unit -> _ = "" [@@bs.obj]

(* translate [`a] to 0, [`b] to 1 *)
type int_expect = < x__ignore: int > Js.t

let int_expect : int_expect = int_test ~x__ignore:`a ()

external int_test2 : ?x__ignore:([`a | `b][@bs.int]) -> unit -> _ = ""
  [@@bs.obj]

type int_expect2 = < x__ignore: int Js.Undefined.t > Js.t

let int_expect2 : int_expect2 = int_test2 ~x__ignore:`a ()

external int_test3 : ?x__ignore:([`a[@bs.as 2] | `b][@bs.int]) -> unit -> _
  = ""
  [@@bs.obj]

let int_expects : int_expect2 list =
  [int_test3 (); int_test3 ~x__ignore:`a (); int_test3 ~x__ignore:`b ()]

external ice :
  flavour:([`vanilla | `chocolate][@bs.string]) -> num:int -> unit -> _ = ""
  [@@bs.obj]

let mk_ice : < flavour: string ; num: int > Js.t =
  ice ~flavour:`vanilla ~num:3 ()

external ice2 :
  ?flavour:([`vanilla | `chocolate][@bs.string]) -> num:int -> unit -> _ = ""
  [@@bs.obj]

let my_ice2 : < flavour: string Js.Undefined.t ; num: int > Js.t =
  ice2 ~flavour:`vanilla ~num:1 ()

let my_ice3 : < flavour: string Js.Undefined.t ; num: int > Js.t =
  ice2 ~num:2 ()

external mk4 : x__ignore:([`a | `b][@bs.ignore]) -> y:int -> unit -> _ = ""
  [@@bs.obj]

let v_mk4 : < y: int > Js.t = mk4 ~x__ignore:`a ~y:3 ()

external mk5 : x:unit -> y:int -> unit -> _ = "" [@@bs.obj]

let v_mk5 : < x: unit ; y: int > Js.t = mk5 ~x:() ~y:3 ()

external mk6 : ?x:unit -> y:int -> unit -> _ = "" [@@bs.obj]

let v_mk6 : < x: unit Js.Undefined.t ; y: int > Js.t = mk6 ~y:3 ()
let v_mk6_1 = mk6 ~x:() ~y:3 ()

type mk

external mk : ?x__ignore:([`a | `b][@bs.int]) -> unit -> _ = "" [@@bs.obj]

(* TODO: fix me *)
let mk_u : < x__ignore: int Js.Undefined.t > Js.t = mk ~x__ignore:`a ()

external mk7 : ?x:([`a | `b][@bs.ignore]) -> y:int -> unit -> _ = "" [@@bs.obj]

let v_mk7 : < y: int > Js.t list =
  [mk7 ~x:`a ~y:3 (); mk7 ~x:`b ~y:2 (); mk7 ~y:2 ()]

external again : ?x__ignore:([`a | `b][@bs.string]) -> int -> unit = "again"
  [@@bs.val]

let () =
  again ~x__ignore:`a 3 ;
  again 3 ;
  again ?x__ignore:None 3 ;
  again ?x__ignore:(ignore 3 ; None) 3

external again2 : x__ignore:([`a | `b][@bs.string]) -> int -> unit = "again2"
  [@@bs.val]

let () = again2 ~x__ignore:`a 3

external again3 : x__ignore:([`a | `b][@bs.ignore]) -> int -> unit = "again3"
  [@@bs.val]

let () =
  again3 ~x__ignore:`a 3 ;
  again3 ~x__ignore:`b 2

external again4 : ?x:unit -> y:unit -> int -> unit -> unit = "again4"
  [@@bs.val]

let side_effect = ref 0

let () =
  again4 ~y:() __LINE__ () ;
  again4 ?x:None ~y:() __LINE__ () ;
  again4 ?x:(Some ()) ~y:() __LINE__ () ;
  again4 ~x:() ~y:() __LINE__ () ;
  again4 ~y:() __LINE__ () ;
  again4 ~x:(incr side_effect ; ()) ~y:() __LINE__ () ;
  again4 ~x:(incr side_effect ; ()) ~y:(decr side_effect ; ()) __LINE__ () ;
  again4 ~y:(decr side_effect ; ()) __LINE__ () ;
  again4 ~x:(incr side_effect) ~y:() __LINE__ ()

(* external again5 : ?x__ignore:([`a of unit -> int | `b of string -> int ]
   [@bs.string]) *)
(* -> int -> unit = "" [@@bs.val] *)

(* let v = again5 3 *)

File: jscomp/test/gpr_1072.js
'use strict';


var u = {
  y: 3
};

var v_ice_cream3_0 = {
  flavor: "vanilla",
  num: 3
};

var v_ice_cream3_1 = {
  hd: {
    flavor: "x",
    num: 3
  },
  tl: {
    hd: {
      flavor: "vanilla",
      num: 3
    },
    tl: /* [] */0
  }
};

var v_ice_cream3 = {
  hd: v_ice_cream3_0,
  tl: v_ice_cream3_1
};

var v_ice_cream4_0 = {
  flavor: "vanilla",
  num: 3
};

var v_ice_cream4_1 = {
  hd: {
    flavor: "x",
    num: 3
  },
  tl: /* [] */0
};

var v_ice_cream4 = {
  hd: v_ice_cream4_0,
  tl: v_ice_cream4_1
};

var vv = {
  x: 3
};

var int_expect = {
  x: 0
};

var int_expect2 = {
  x: 0
};

var int_expects_0 = {};

var int_expects_1 = {
  hd: {
    x: 2
  },
  tl: {
    hd: {
      x: 3
    },
    tl: /* [] */0
  }
};

var int_expects = {
  hd: int_expects_0,
  tl: int_expects_1
};

var mk_ice = {
  flavour: "vanilla",
  num: 3
};

var my_ice2 = {
  flavour: "vanilla",
  num: 1
};

var my_ice3 = {
  num: 2
};

var v_mk4 = {
  y: 3
};

var v_mk5 = {
  x: undefined,
  y: 3
};

var v_mk6 = {
  y: 3
};

var v_mk6_1 = {
  x: undefined,
  y: 3
};

var mk_u = {
  x: 0
};

var v_mk7_0 = {
  y: 3
};

var v_mk7_1 = {
  hd: {
    y: 2
  },
  tl: {
    hd: {
      y: 2
    },
    tl: /* [] */0
  }
};

var v_mk7 = {
  hd: v_mk7_0,
  tl: v_mk7_1
};

again("a", 3);

again(undefined, 3);

again(undefined, 3);

again(undefined, 3);

again2("a", 3);

again3(3);

again3(2);

var side_effect = {
  contents: 0
};

again4(undefined, undefined, 166);

again4(undefined, undefined, 167);

again4(undefined, undefined, 168);

again4(undefined, undefined, 169);

again4(undefined, undefined, 170);

again4((side_effect.contents = side_effect.contents + 1 | 0, undefined), undefined, 171);

again4((side_effect.contents = side_effect.contents + 1 | 0, undefined), (side_effect.contents = side_effect.contents - 1 | 0, undefined), 172);

again4(undefined, (side_effect.contents = side_effect.contents - 1 | 0, undefined), 173);

again4((side_effect.contents = side_effect.contents + 1 | 0, undefined), undefined, 174);

exports.u = u;
exports.v_ice_cream3 = v_ice_cream3;
exports.v_ice_cream4 = v_ice_cream4;
exports.vv = vv;
exports.int_expect = int_expect;
exports.int_expect2 = int_expect2;
exports.int_expects = int_expects;
exports.mk_ice = mk_ice;
exports.my_ice2 = my_ice2;
exports.my_ice3 = my_ice3;
exports.v_mk4 = v_mk4;
exports.v_mk5 = v_mk5;
exports.v_mk6 = v_mk6;
exports.v_mk6_1 = v_mk6_1;
exports.mk_u = mk_u;
exports.v_mk7 = v_mk7;
exports.side_effect = side_effect;
/*  Not a pure module */
